File: src/lib/windowing.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# vim:fenc=utf-8

import numpy as np


def windowing(arr, window_len):
    """Perform time series windowing transformation."""
    num_cases = len(arr) - window_len + 1
    cases = [arr[i:(i + window_len)] for i in range(num_cases)]
    return np.array(cases)

File: src/lib/windowing.R
windowing <- function(x, windowLen) {
  # Summary: Transform a time series data into a set of cases.
  # Reference: Autoregressive Tree Models for Time-Series Analysis (C. Meek, 2002)
  #
  # Arguments:
  #   x: A numeric vector or time series.
  #   windowLen: The length of windowing transformation (must be an integer >= 2).
  #
  # Returns:
  #   A list of cases in matrix format
  
  # Check precondition
  if (!(windowLen >= 2)) {
    stop("The value of windowLen must be an integer >= 2.")
  }
  
  tsData <- ts(x)
  cases <- c()
  
  numCases <- length(tsData) - windowLen + 1
  for(i in 1:numCases)
  {
    cases <- rbind(cases, window(tsData, i, i + windowLen - 1)[1:windowLen])
  }

  return(cases)
}